common/src/main/java/com/redhat/lightblue/rest/cors/CorsConfiguration.java

```diff
@@ -0,0 +1,518 @@
+    private final List<String> allowedOrigins;
+    private final List<String> allowedMethods;
+    private final List<String> allowedHeaders;
+    private final List<String> exposedHeaders;
+    private final int preflightMaxAge;
+    private final boolean allowCredentials;
+    private final boolean enableLogging;
+
+    /**
+     * Chain setters together to configure a {@link com.redhat.lightblue.rest.cors.CorsConfiguration}
+     * object.
+     *
+     * <p>To read values from a JSON configuration file, use {@link #fromJson(java.nio.file.Path)}.
+     *
+     * <p>To understand what these configurations mean in more depth, see
+     * <a href="https://developer.mozilla.org/en-US/docs/Web/HTTP/Access_control_CORS">HTTP
+     * access control documentation on MDN.</a>
+     */
+    public static class Builder {
+        static final List<String> DEFAULT_URL_PATTERNS = unmodifiableList(Arrays.asList("/*"));
+        static final List<String> DEFAULT_ALLOWED_ORIGINS = unmodifiableList(Arrays.asList("*"));
+        static final List<String> DEFAULT_ALLOWED_METHODS = unmodifiableList(Arrays.asList("GET",
+                "PUT", "POST", "HEAD", "OPTIONS"));;
+        static final List<String> DEFAULT_ALLOWED_HEADERS = unmodifiableList(Arrays.asList("Origin",
+                "Accept", "X-Requested-With", "Content-Type", "Access-Control-Request-Method",
+                "Access-Control-Request-Headers"));
+        static final List<String> DEFAULT_EXPOSED_HEADERS = Collections.emptyList();
+        static final int DEFAULT_PREFLIGHT_MAX_AGE = 1800;
+        static final boolean DEFAULT_ALLOW_CREDENTIALS = true;
+        static final boolean DEFAULT_ENABLE_LOGGING = false;
+
+        private List<String> urlPatterns = DEFAULT_URL_PATTERNS;
+        private List<String> allowedOrigins = DEFAULT_ALLOWED_ORIGINS;
+        private List<String> allowedMethods = DEFAULT_ALLOWED_METHODS;
+        private List<String> allowedHeaders = DEFAULT_ALLOWED_HEADERS;
+        private List<String> exposedHeaders = DEFAULT_EXPOSED_HEADERS;
+        private int preflightMaxAge = DEFAULT_PREFLIGHT_MAX_AGE;
+        private boolean allowCredentials = DEFAULT_ALLOW_CREDENTIALS;
+        private boolean enableLogging = DEFAULT_ENABLE_LOGGING;
+
+        /**
+         * Sets the URL patterns to use with a copy of the contents from the provided array,
+         * preventing later modifications to the array from affecting the
+         * {@link com.redhat.lightblue.rest.cors.CorsConfiguration}.
+         *
+         * <p>URL matching rules follow that of servlet and filter mappings' URL patterns.
+         *
+         * <p>Defaults to "/*".
+         */
+        public Builder urlPatterns(String... patterns) {
+            Objects.requireNonNull(patterns, "patterns");
+
+            String[] copy = new String[patterns.length];
+            System.arraycopy(patterns, 0, copy, 0, patterns.length);
+
+            this.urlPatterns = unmodifiableList(Arrays.asList(copy));
```

Comment:
The backing list (and its backing array) can still change, unmodifiableList is just a view onto a list that prevents further modifications. An unmodifiable list is unfortunately not immutable so this is why I'm kind of going through hoops here to be clean about it.